csp_billing_adapter/bill_utils.py

```diff
@@ -212,6 +214,86 @@ def get_volume_dimensions(
+    usage_metric: str,
+    usage: int,
+    metric_dimensions: dict,
+    billed_dimensions: dict
+) -> None:
+    """
+    For the metric specified by 'usage_metric', with the given
+    'usage' value, determine the appropiate tiered dimensions
+    that correspond to that usage level, as specifed by the
+    'metric_dimensions' hash, and update the 'billed_dimensions'
+    hash with entries mapping those dimensions names to the
+    appropriate usage amounts.
+
+    Note that, unlike the volume dimensions case, where all usage
+    is reported under the single dimension into which the usage
+    level falls, in the case of tiered dimensions, the usage
+    amount is allocated in an effort to fill each dimension, with
+    the excess overflowing into the next dimension, meaning that
+    multiple dimensions can, potentially, be reported.
+
+    :param usage_metric: The metric being processed.
+    :param usage: The calculated usage for the specified metric.
+    :param metric_dimensions:
+        The hash specifying the tiered dimensions for the usage
+        metric.
+    :param billed_dimensions:
+        The hash that will be updated with an entry mapping the
+        determined tiered dimension names to the appropriate
+        usage values for that dimension.
+    """
+    unassigned_usage = usage
+    for dimension in metric_dimensions:
+        # handle a missing min, as well as allowing usage and min
+        # being 0 to generate a billing dimension entry with the
+        # appropriate dimension name and a value of 0.
+        if 'min' in dimension and usage < dimension['min']:
+            log.debug("Skipping as usage < min")
+            continue
+
+        # if configured min is not specified or 0, use 1 as the
+        # min value since our later range calculations are for
+        # inclusive range specifications.
+        dim_min = max(dimension.get('min', 0), 1)
+
+        # use the configured max or usage if none is specified
+        dim_max = dimension.get('max', usage)
+
+        # the usage value for this dimension is calculated by
+        # subtracting the dim_min from the lesser of usage or
+        # dim_max, and then add 1 because the range is inclusive
+        dim_usage = (min(usage, dim_max) - dim_min) + 1
+
+        billed_dimensions[dimension['dimension']] = dim_usage
+
+        # subtract assigned usage
+        unassigned_usage -= dim_usage
+
+        log.debug(
+            "Adding '%s'=%d to billing_dimensions",
+            dimension['dimension'],
+            dim_usage
+        )
+
+    # unassigned usage indicates that there is a missing tiered
+    # dimension specification in the config
+    if unassigned_usage:
+        log.error(
+            "Failed to find a matching dimension entry for %s=%d, "
+            "%d unassigned",
+            usage_metric,
+            usage,
+            unassigned_usage
+        )
+        raise MissingTieredDimensionError(
+            usage_metric,
+            usage
```

Comment:
In a tiered model the last dimension in the config should not have a `max`, meaning whatever is left over after the next to last dimension is satisfied gets billed to the last dimension. This is also implemented on lines 263 and 268. As such I'd rather see us compensate for an incorrect config in a different way and add the leftover usage to the last calculation vs. triggering an error. The error that is implemented here to be triggered is really something that should be raised during config validation as "LastDimeansionInTieredModelHasMaxError". If I recall correctly the implementation of config validation was postponed, if that recollection is correct it now works against us.

Reply:
So this check is intended to catch the case where someone created a gap in the dimensions settings, e.g they specified (min, max) settings that max of the previous dimension is not (min-1) of the current setting. In such a case, we will allocate all of the "remaining" units, above the dimensions min setting, to the final dimension's bucket, but if there was a gap anywhere, the units associated with the gap's range cannot be allocated...

If we instead just specified tiered dimensions with just a bucket size, then in that case we would allocate the corresponding amount to each bucket, and whatever was left over would be allocated to the last bucket.

Note that this implementation will still run into issues if dimensions other than the first one fail to specify a min value, as that will effectively turn the max setting into a bucket size...

So yeah it might not be a bad idea to add some config validation support, at least for dimension specification sanity.